drive/src/drive/contract.rs

```diff
@@ -0,0 +1,599 @@
+                        apply,
+                        insert_operations,
+                    )?;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn apply_contract_cbor(
+        &self,
+        contract_cbor: Vec<u8>,
+        contract_id: Option<[u8; 32]>,
+        block_time: f64,
+        apply: bool,
+        transaction: TransactionArg,
+    ) -> Result<(i64, u64), Error> {
+        // first we need to deserialize the contract
+        let contract = Contract::from_cbor(&contract_cbor, contract_id)?;
+        self.apply_contract(&contract, contract_cbor, block_time, apply, transaction)
+    }
+
+    pub fn get_contract(
+        &self,
+        contract_id: [u8; 32],
+        transaction: TransactionArg,
+    ) -> Result<Option<Arc<Contract>>, Error> {
+        let cached_contracts = self.cached_contracts.borrow();
+        match cached_contracts.get(&contract_id) {
+            None => self.fetch_contract(contract_id, transaction),
+            Some(contract) => {
+                let contract_ref = Arc::clone(&contract);
+                Ok(Some(contract_ref))
+            }
+        }
+    }
+
+    pub fn get_cached_contract(
+        &self,
+        contract_id: [u8; 32],
+    ) -> Result<Option<Arc<Contract>>, Error> {
+        let cached_contracts = self.cached_contracts.borrow();
+        match cached_contracts.get(&contract_id) {
+            None => Ok(None),
+            Some(contract) => {
+                let contract_ref = Arc::clone(&contract);
+                Ok(Some(contract_ref))
+            }
+        }
+    }
+
+    pub fn fetch_contract(
+        &self,
+        contract_id: [u8; 32],
+        transaction: TransactionArg,
+    ) -> Result<Option<Arc<Contract>>, Error> {
+        let stored_element = self
+            .grove
+            .get(contract_root_path(&contract_id), &[0], transaction)?;
+        if let Element::Item(stored_contract_bytes) = stored_element {
+            let contract = Arc::new(Contract::from_cbor(&stored_contract_bytes, None)?);
+            let cached_contracts = self.cached_contracts.borrow();
+            cached_contracts.insert(contract_id, Arc::clone(&contract));
+            Ok(Some(Arc::clone(&contract)))
+        } else {
+            Err(Error::Drive(DriveError::CorruptedContractPath(
+                "contract path did not refer to a contract element",
+            )))
+        }
+    }
+
+    pub fn apply_contract(
+        &self,
+        contract: &Contract,
+        contract_serialization: Vec<u8>,
+        block_time: f64,
+        apply: bool,
+        transaction: TransactionArg,
+    ) -> Result<(i64, u64), Error> {
+        let mut query_operations: Vec<QueryOperation> = vec![];
+        let mut insert_operations: Vec<InsertOperation> = vec![];
+
+        // overlying structure
+        let mut already_exists = false;
+        let mut original_contract_stored_data = vec![];
+
+        if let Ok(Some(stored_element)) = self.grove_get(
```

Comment:
It will always gonna do insert in case of application, right? But I feel like a contract update can be havier.